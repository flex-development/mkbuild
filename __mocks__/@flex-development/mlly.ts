/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
