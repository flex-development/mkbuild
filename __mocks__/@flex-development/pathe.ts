/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
