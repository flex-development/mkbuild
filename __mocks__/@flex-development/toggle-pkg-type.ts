/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
