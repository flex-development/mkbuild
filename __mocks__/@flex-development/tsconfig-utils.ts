/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
