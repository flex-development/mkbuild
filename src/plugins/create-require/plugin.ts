/**
 * @file Plugins - create-require
 * @module mkbuild/plugins/create-require/plugin
 */

import { constant, define, join, regexp } from '@flex-development/tutils'
import {
  transform,
  type BuildOptions,
  type BuildResult,
  type OutputFile,
  type Plugin,
  type PluginBuild
} from 'esbuild'
import util from 'node:util'

/**
 * Plugin-specific build options.
 *
 * @internal
 */
type SpecificOptions = { metafile: true; write: false }

/**
 * Returns a plugin that defines the `require` function in ESM bundles.
 *
 * When outputting ESM, any `require` calls will be replaced with esbuild's
 * `__require` shim, since [`require` is not defined in ESM environments][1].
 *
 * The shim first checks if `require` is polyfilled. If it is, the shim passes
 * the call to `require.apply`. If `require` isn't defined, an error will be
 * thrown: `Dynamic require of "' + x + '" is not supported`.
 *
 * This is particularly problematic when bundling, especially when targeting
 * node ([`platform === 'node'`][2]), since [built-in modules][3] are marked
 * as [`external`][4] by esbuild automatically. `require`ing these modules, as
 * well as other externals, will generate a runtime error. This is described
 * in [`evanw/esbuild#1921`][5].
 *
 * To circumvent this issue, output files containing the shim will have a
 * snippet insert that defines `require` using [`module.createRequire`][6].
 *
 * [1]: https://nodejs.org/api/esm.html#no-require-exports-or-moduleexports
 * [2]: https://esbuild.github.io/api/#platform
 * [3]: https://nodejs.org/api/esm.html#builtin-modules
 * [4]: https://esbuild.github.io/api/#external
 * [5]: https://github.com/evanw/esbuild/issues/1921
 * [6]: https://nodejs.org/api/module.html#modulecreaterequirefilename
 *
 * @return {Plugin} `create-require` plugin
 */
const plugin = (): Plugin => {
  /**
   * Inserts a `require` function definition into ESM bundles.
   *
   * [1]: https://esbuild.github.io/plugins
   * [2]: https://esbuild.github.io/api/#build-api
   *
   * @param {PluginBuild} build - [esbuild plugin api][1]
   * @param {BuildOptions} build.initialOptions - [esbuild build api][2] options
   * @param {PluginBuild['onEnd']} build.onEnd - Build end callback
   * @return {Promise<void>} Nothing when complete
   */
  const setup = async ({
    initialOptions,
    onEnd
  }: PluginBuild): Promise<void> => {
    const {
      absWorkingDir = process.cwd(),
      banner: { js: banner = '' } = {},
      bundle,
      format = 'esm',
      metafile,
      minify = false,
      minifySyntax,
      minifyWhitespace,
      platform = 'node',
      target = [],
      write
    } = initialOptions

    // do nothing if bundling is not enabled
    if (!bundle) return void bundle

    // do nothing if not creating esm bundle
    if (format !== 'esm') return void format

    // esbuild write must be disabled to access result.outputFiles
    if (write) throw new Error('write must be disabled')

    // metafile required to get output metadata
    if (!metafile) throw new Error('metafile required')

    /**
     * Code snippet that defines `require` using [`module.createRequire`][1].
     *
     * [1]: https://nodejs.org/api/module.html#modulecreaterequirefilename
     *
     * @const {string} snippet
     */
    const snippet: string = join(
      [
        'import { createRequire as __createRequire } from \'node:module\'',
        'const require = __createRequire(import.meta.url)'
      ],
      '\n'
    )

    // transform snippet to re-add banner and minify
    let { code } = await transform(snippet, {
      banner,
      minifySyntax: minifySyntax ?? minify,
      minifyWhitespace: minifyWhitespace ?? minify,
      platform,
      target
    })

    // remove new line from end of code snippet if output should be minified
    if (minify || minifyWhitespace) code = code.replace(/\n$/, '')

    return void onEnd((result: BuildResult<SpecificOptions>): void => {
      /**
       * Regex used to deduce if an output file includes the `__require` shim.
       *
       * @const {RegExp} filter
       */
      const filter: RegExp = /Dynamic require of ".*" is not supported/m

      // insert require function definitions
      result.outputFiles = result.outputFiles.map((output: OutputFile) => {
        // do nothing if output file does not contain shim
        if (!filter.test(output.text)) return output

        // get hashbang
        const [hashbang = ''] = /^#!.+\n/.exec(output.text) ?? []

        /**
         * {@linkcode output.text} copy.
         *
         * @var {string} text
         */
        let text: string = output.text.replace(hashbang, '')

        // remove banner
        if (banner) text = text.replace(new RegExp(regexp(banner) + '\n?'), '')

        // redefine output text to insert require function definition
        define(output, 'text', { get: constant(hashbang + code + text) })

        // reset output file contents
        output.contents = new util.TextEncoder().encode(output.text)

        /**
         * Relative path to output file.
         *
         * **Note**: Relative to {@linkcode absWorkingDir}.
         *
         * @const {string} outfile
         */
        const outfile: string = output.path
          .replace(absWorkingDir, '')
          .replace(/^\//, '')

        /**
         * Output contents size including `require` function definition.
         *
         * @const {number} bytes
         */
        const bytes: number = Buffer.byteLength(output.contents)

        // reset output file size
        result.metafile.outputs[outfile]!.bytes = bytes

        return output
      })
    })
  }

  return { name: 'create-require', setup }
}

export { plugin as default }
