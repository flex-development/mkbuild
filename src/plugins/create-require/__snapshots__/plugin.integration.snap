// Vitest Snapshot v1

exports[`integration:plugins/create-require > esbuild > should insert require definition 1`] = `
"/**
 * @file Virtual File System
 * @module vfs
 * @see https://github.com/streamich/memfs
 */
import { createRequire as __createRequire } from \\"node:module\\";
const require = __createRequire(import.meta.url);
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __require = /* @__PURE__ */ ((x) => typeof require !== \\"undefined\\" ? require : typeof Proxy !== \\"undefined\\" ? new Proxy(x, {
  get: (a, b) => (typeof require !== \\"undefined\\" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== \\"undefined\\")
    return require.apply(this, arguments);
  throw new Error('Dynamic require of \\"' + x + '\\" is not supported');
});
"
`;

exports[`integration:plugins/create-require > esbuild > should insert require definition into minified output 1`] = `"import{createRequire as __createRequire}from\\"node:module\\";const require=__createRequire(import.meta.url);var __create=Object.create;var __defProp=Object.defineProperty;var __getOwnPropDesc=Object.getOwnPropertyDescriptor;var __getOwnPropNames=Object.getOwnPropertyNames;var __getProtoOf=Object.getPrototypeOf;var __hasOwnProp=Object.prototype.hasOwnProperty;var __require=(x=>typeof require!==\\"undefined\\"?require:typeof Proxy!==\\"undefined\\"?new Proxy(x,{get:(a,b)=>(typeof require!==\\"undefined\\"?require:a)[b]}):x)(function(x){if(typeof require!==\\"undefined\\")return require.apply(this,arguments);throw new Error('Dynamic require of \\"'+x+'\\" is not supported')});"`;

exports[`integration:plugins/create-require > esbuild > should skip output files without __require shim 1`] = `
"// src/max-value.ts
var max_value_default = 2 ** 31 - 1;

// src/min-value.ts
var min_value_default = (-2) ** 31;

// src/reverse.mts
function reverse(x) {
  if (x >= -9 && x <= 9)
    return x;
  const s = x.toString();
  const r = Number.parseInt([...s, x < 0 ? \\"-\\" : \\"\\"].reverse().join(\\"\\"));
  return r < min_value_default || r > max_value_default ? 0 : r;
}
var reverse_default = reverse;
export {
  reverse_default as default
};
"
`;
